Fixes/more type hints 004

Quix0r opened the pull request:
The next batch of changes:
- type-hints added
- fixed returned type-hint (`void` was no longer valid there)
- added some documentation
- added `default:` block for unknown relation

Quix0r committed:
Changes:
- added more type-hints
- added missing documentation

Quix0r committed:
Changes:
- added type-hints
- added documentation

Quix0r committed:
Changes
- added type-hints
- added some documentation
- added default: block for unknown 'rel' value
- changed return type void to empty array

Quix0r committed:
Ops, forgot to rename these, too.

Quix0r committed:
$data can turn into bool here

Quix0r committed:
Changes:
- addHubLink()'s 2nd parameter is DOMElement
- added more type-hints
- converted double-quotes to single

Quix0r committed:
Changes:
- added more type-hints
- updated documentation
- changed double-quotes to single

Quix0r committed:
Return empty array on error (I hope it works this way).

Quix0r committed:
Changes:
- added more type-hints
- added some documentation

Quix0r committed:
Changes:
- added more type-hints
- added missing documentation
- Email::send() now returns what mail() returns (bool)

MrPetovan commented on src/BaseModule.php:
```diff
@@ -279,12 +284,12 @@ protected function checkDefaults(array $defaults, array $input): array
-	 * @param mixed $minimal_value 
-	 * @param mixed $maximum_value 
-	 * @return mixed 
+	 * @param array $input Input viels
```
```suggestion
	 * @param array $input Input fields
```

MrPetovan commented on src/Model/Tag.php:
```diff
@@ -73,11 +73,12 @@ class Tag
-	 * @param string  $name
-	 * @param string  $url
-	 * @param integer $target
+	 * @param integer $uriid URI id
+	 * @param integer $type Type
+	 * @param string  $name Name
+	 * @param string  $url URL
+	 * @param integer $target Target (default: null)
```
There's no point repeating the parameter name in the documentation. If you don't have anything more to add, please leave it as is. It is fine for now, but please be mindful in the future.

Quix0r replied:
Okay, I have expanded it a bit.

MrPetovan commented on src/Network/Probe.php:
```diff
@@ -143,27 +141,26 @@ private static function ownHost($host)
 
 	/**
-	 * Probes for webfinger path via "host-meta"
+	 * Probes for webfinger path via 'host-meta'
```
So, here's the thing. For strings PHP code accepts single and double quotes (and backticks too) but in regular language, quoted text only accepts double-quotes (or double chevrons in French for example). While replacing double-quotes by single quotes in the PHP code makes sense, it doesn't as much in English comments.
```suggestion
	 * Probes for webfinger path via "host-meta"
```

MrPetovan commented on src/Network/Probe.php:
```diff
@@ -1137,73 +1131,72 @@ public static function profile($profile_link)
+			} elseif (($link['rel'] == 'diaspora-public-key') && !empty($link['href'])) {
+				$data['pubkey'] = base64_decode($link['href']);
+
+				//if (strstr($data['pubkey'], 'RSA ') || ($link['type'] == 'RSA'))
```
Please remove this commented out code.
```suggestion
```

MrPetovan commented on src/Network/Probe.php:
```diff
@@ -1299,41 +1291,41 @@ private static function pollHcard($hcard_url, $data, $dfrn = false)
 			if ($search->length > 0) {
 				foreach ($search as $link) {
-					//$data["request"] = $search->item(0)->nodeValue;
+					//$data['request'] = $search->item(0)->nodeValue;
```
Please remove this commented out code.
```suggestion
```

MrPetovan commented on src/Network/Probe.php:
```diff
@@ -1299,41 +1291,41 @@ private static function pollHcard($hcard_url, $data, $dfrn = false)
-			// Older Friendica versions had used the "uid" field differently than newer versions
-			if (!empty($data["nick"]) && !empty($data["guid"]) && ($data["nick"] == $data["guid"])) {
-				unset($data["guid"]);
+			// Older Friendica versions had used the 'uid' field differently than newer versions
```
```suggestion
			// Older Friendica versions had used the "uid" field differently than newer versions
```

MrPetovan commented on src/Network/Probe.php:
```diff
@@ -1345,64 +1337,63 @@ private static function pollHcard($hcard_url, $data, $dfrn = false)
+			} elseif (($link['rel'] == 'diaspora-public-key') && !empty($link['href'])) {
+				$data['pubkey'] = base64_decode($link['href']);
+
+				//if (strstr($data['pubkey'], 'RSA ') || ($link['type'] == 'RSA'))
```
```suggestion
```

MrPetovan commented on src/Network/Probe.php:
```diff
@@ -1412,23 +1403,23 @@ private static function diaspora($webfinger)
-			// We have to overwrite the detected value for "notify" since Hubzilla doesn't send it
-			$data["notify"] = $data["baseurl"] . "/receive/users/" . $data["guid"];
-			$data["batch"]  = $data["baseurl"] . "/receive/public";
+			// We have to overwrite the detected value for 'notify' since Hubzilla doesn't send it
```
```suggestion
			// We have to overwrite the detected value for "notify" since Hubzilla doesn't send it
```

MrPetovan commented on src/Network/Probe.php:
```diff
@@ -1991,9 +1977,7 @@ public static function getLastUpdate(array $data)
-	 * @return string last activity
-	 *
-	 * @return bool 'true' if update was successful or the server was unreachable
+	 * @return string last activity or bool 'true' if update was successful or the server was unreachable
```
```suggestion
	 * @return string last activity or true if update was successful or the server was unreachable
```

MrPetovan commented on src/Object/Image.php:
```diff
@@ -650,11 +674,14 @@ public function crop($max, $x, $y, $w, $h)
-		return $this->asString();
+	public function __toString(): string
+	{
+		return (string) $this->asString();
```
Of course `asString` doesn't return just a string. 😓

(Nothing you should do about it now, I'm just venting)

Quix0r replied:
`asString()` may return `false` on failure so I better cast it to `string` as `__toString()` will always return `string` ...

MrPetovan replied:
You did good, the original code is 💩

MrPetovan commented on src/Protocol/DFRN.php:
```diff
@@ -186,17 +188,17 @@ public static function itemFeed(int $uri_id, int $uid, bool $conversation = fals
 
-			//$root = self::addHeader($doc, $owner, "dfrn:owner", "", false);
+			//$root = self::addHeader($doc, $owner, 'dfrn:owner', '', false);
 
```
```suggestion
```

MrPetovan commented on src/Protocol/DFRN.php:
```diff
@@ -1639,11 +1659,10 @@ private static function doPoke(array $item, array $importer)
 	private static function processVerbs(int $entrytype, array $importer, array &$item, bool &$is_like)
 	{
-		Logger::info("Process verb ".$item["verb"]." and object-type ".$item["object-type"]." for entrytype ".$entrytype);
+		Logger::info("Process verb " . $item['verb'] . " and object-type " . $item['object-type'] . " for entrytype " . $entrytype);
```
```suggestion
		Logger::info('Process verb ' . $item['verb'] . ' and object-type ' . $item['object-type'] . ' for entrytype ' . $entrytype);
```

MrPetovan commented on src/Protocol/Email.php:
```diff
@@ -398,52 +402,70 @@ public static function send($addr, $subject, $headers, $item)
 	/**
-	 * @param string $iri string
-	 * @return string
+	 * Convert iri (?) to message id
```
From the uses of this method I believe `iri` is `Item URI`.

Quix0r replied:
I updated the documentation here + renamed `$iri` to $itemUri` which is more clear.

MrPetovan commented on src/Protocol/Feed.php:
```diff
@@ -66,7 +67,7 @@ public static function import(string $xml, array $importer = [], array $contact
 			Logger::info("Test Atom/RSS feed");
 		} else {
-			Logger::info("Import Atom/RSS feed '" . $contact["name"] . "' (Contact " . $contact["id"] . ") for user " . $importer["uid"]);
+			Logger::info("Import Atom/RSS feed '" . $contact['name'] . "' (Contact " . $contact['id'] . ") for user " . $importer['uid']);
```
```suggestion
			Logger::info('Import Atom/RSS feed "' . $contact['name'] . '" (Contact ' . $contact['id'] . ') for user ' . $importer['uid']);
```

MrPetovan commented on src/Protocol/Feed.php:
```diff
@@ -378,7 +379,7 @@ public static function import(string $xml, array $importer = [], array $contact
 					// Use the creation date when the post had been stored. It can happen this date changes in the feed.
 					$creation_dates[] = $previous['created'];
-					Logger::info("Item with uri " . $item["uri"] . " for user " . $importer["uid"] . " already existed under id " . $previous["id"]);
+					Logger::info("Item with uri " . $item['uri'] . " for user " . $importer['uid'] . " already existed under id " . $previous['id']);
```
```suggestion
					Logger::info('Item with uri ' . $item['uri'] . ' for user ' . $importer['uid'] . ' already existed under id ' . $previous['id']);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -354,34 +372,34 @@ private static function process($xml, array $importer, array &$contact = null, &
+					if ($hub_attribute->name == 'href') {
 						$hub = $hub_attribute->textContent;
-						Logger::info("Found hub ".$hub);
+						Logger::info("Found hub ", ['hub' => $hub]);
```
```suggestion
						Logger::info('Found hub ', ['hub' => $hub]);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -427,67 +445,67 @@ private static function process($xml, array $importer, array &$contact = null, &
-				$orig_uri = $xpath->query("activity:object/atom:id", $entry)->item(0)->nodeValue;
+			if ($item['verb'] == Activity::FAVORITE) {
+				$orig_uri = $xpath->query('activity:object/atom:id', $entry)->item(0)->nodeValue;
 				Logger::notice("Favorite", ['uri' => $orig_uri, 'item' => $item]);
```
```suggestion
				Logger::notice('Favorite', ['uri' => $orig_uri, 'item' => $item]);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -179,41 +191,41 @@ private static function fetchAuthor(DOMXPath $xpath, $context, array $importer,
 
-			if (!empty($author["author-avatar"]) && ($author["author-avatar"] != $current['avatar'])) {
+			if (!empty($author['author-avatar']) && ($author['author-avatar'] != $current['avatar'])) {
 				Logger::info("Update profile picture for contact ".$contact["id"]);
```
```suggestion
				Logger::info('Update profile picture for contact ' . $contact['id']);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -427,67 +445,67 @@ private static function process($xml, array $importer, array &$contact = null, &
-			if (in_array($item["verb"], [Activity::O_UNFAVOURITE, Activity::UNFAVORITE])) {
+			if (in_array($item['verb'], [Activity::O_UNFAVOURITE, Activity::UNFAVORITE])) {
 				// Ignore "Unfavorite" message
 				Logger::info("Ignore unfavorite message ", ['item' => $item]);
```
```suggestion
				Logger::info('Ignore unfavorite message ', ['item' => $item]);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -427,67 +445,67 @@ private static function process($xml, array $importer, array &$contact = null, &
-			if ($item["verb"] == Activity::JOIN) {
+			if ($item['verb'] == Activity::JOIN) {
 				// ignore "Join" messages
 				Logger::info("Ignore join message ", ['item' => $item]);
```
```suggestion
				Logger::info('Ignore join message ', ['item' => $item]);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -427,67 +445,67 @@ private static function process($xml, array $importer, array &$contact = null, &
-			if ($item["verb"] == "http://mastodon.social/schema/1.0/block") {
+			if ($item['verb'] == 'http://mastodon.social/schema/1.0/block') {
 				// ignore mastodon "block" messages
 				Logger::info("Ignore block message ", ['item' => $item]);
```
```suggestion
				Logger::info('Ignore block message ', ['item' => $item]);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -427,67 +445,67 @@ private static function process($xml, array $importer, array &$contact = null, &
-			if (!in_array($item["verb"], [Activity::POST, Activity::LIKE, Activity::SHARE])) {
-				Logger::info("Unhandled verb", ['verb' => $item["verb"], 'item' => $item]);
+			if (!in_array($item['verb'], [Activity::POST, Activity::LIKE, Activity::SHARE])) {
+				Logger::info("Unhandled verb", ['verb' => $item['verb'], 'item' => $item]);
```
```suggestion
				Logger::info('Unhandled verb', ['verb' => $item['verb'], 'item' => $item]);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -503,10 +521,10 @@ private static function process($xml, array $importer, array &$contact = null, &
 
 						if ($valid) {
-							Logger::info("Item with uri ".self::$itemlist[0]['uri']." will be imported due to the system settings.");
+							Logger::info("Item with uri " . self::$itemlist[0]['uri'] . " will be imported due to the system settings.");
```
```suggestion
							Logger::info('Item with uri ' . self::$itemlist[0]['uri'] . ' will be imported due to the system settings.');
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -503,10 +521,10 @@ private static function process($xml, array $importer, array &$contact = null, &
 						}
 					} else {
-						Logger::info("Item with uri ".self::$itemlist[0]['uri']." belongs to a contact (".self::$itemlist[0]['contact-id']."). It will be imported.");
+						Logger::info("Item with uri " . self::$itemlist[0]['uri'] . " belongs to a contact (" . self::$itemlist[0]['contact-id'] . "). It will be imported.");
```
```suggestion
						Logger::info('Item with uri ' . self::$itemlist[0]['uri'] . ' belongs to a contact (' . self::$itemlist[0]['contact-id'] . '). It will be imported.');
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -519,7 +537,7 @@ private static function process($xml, array $importer, array &$contact = null, &
 						}
 						if ($valid) {
-							Logger::info("Item with uri ".self::$itemlist[0]['uri']." will be imported since the thread contains posts or shares.");
+							Logger::info("Item with URI " . self::$itemlist[0]['uri'] . " will be imported since the thread contains posts or shares.");
```
```suggestion
							Logger::info('Item with URI ' . self::$itemlist[0]['uri'] . ' will be imported since the thread contains posts or shares.');
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -536,20 +554,20 @@ private static function process($xml, array $importer, array &$contact = null, &
+						$found = Post::exists(['uid' => $importer['uid'], 'uri' => $item['uri']]);
 						if ($found) {
-							Logger::notice("Item with uri ".$item["uri"]." for user ".$importer["uid"]." already exists.");
+							Logger::notice("Item with uri " . $item['uri'] . " for user " . $importer['uid'] . " already exists.");
```
```suggestion
							Logger::notice('Item with uri ' . $item['uri'] . ' for user ' . $importer['uid'] . ' already exists.');
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -536,20 +554,20 @@ private static function process($xml, array $importer, array &$contact = null, &
+							Logger::notice("Item with uri " . $item['uri'] . " for user " . $importer['uid'] . " already exists.");
 						} elseif ($item['contact-id'] < 0) {
-							Logger::notice("Item with uri ".$item["uri"]." is from a blocked contact.");
+							Logger::notice("Item with uri " . $item['uri'] . " is from a blocked contact.");
```
```suggestion
							Logger::notice('Item with uri ' . $item['uri'] . ' is from a blocked contact.');
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -536,20 +554,20 @@ private static function process($xml, array $importer, array &$contact = null, &
 						} else {
 							$ret = Item::insert($item);
-							Logger::info("Item with uri ".$item["uri"]." for user ".$importer["uid"].' stored. Return value: '.$ret);
+							Logger::info("Item with uri " . $item['uri'] . " for user " . $importer['uid'] . " stored. Return value: " . $ret);
```
```suggestion
							Logger::info('Item with uri ' . $item['uri'] . ' for user ' . $importer['uid'] . ' stored. Return value: ' . $ret);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -1002,7 +1027,7 @@ private static function fetchRelated($related, $related_uri, $importer)
 			self::process($xml, $importer, $contact, $hub, $stored, false, Conversation::PULL);
 		} else {
-			Logger::info("XML couldn't be fetched for URI: ".$related_uri." - href: ".$related);
+			Logger::info("XML couldn't be fetched for URI: " . $related_uri . " - href: " . $related);
```
```suggestion
			Logger::info("XML couldn't be fetched for URI: " . $related_uri . ' - href: ' . $related);
```

Quix0r replied:
Instead of "couldn't " I wrote "could not" to avoid escaping the single-quote.

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -1132,12 +1159,16 @@ private static function processLinks($links, array &$item)
 						break;
+
+					default:
+						Logger::warning('Unsupported rel=' . $attribute['rel'] . ',href=' . $attribute['href'] . ',object-type=' . $attribute['object-type']);
```
```suggestion
						Logger::warning('Unsupported rel=' . $attribute['rel'] . ', href=' . $attribute['href'] . ', object-type=' . $attribute['object-type']);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -1148,67 +1179,67 @@ private static function processLinks($links, array &$item)
 			}
 
-			$body = trim($siteinfo["text"])." [url]".$url."[/url]\n[img]".$preview."[/img]";
+			$body = trim($siteinfo['text']) . " [url]" . $url . "[/url]\n[img]" . $preview . "[/img]";
```
```suggestion
			$body = trim($siteinfo['text']) . ' [url]' . $url . "[/url]\n[img]" . $preview . '[/img]';
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -1531,29 +1584,28 @@ private static function entry(DOMDocument $doc, array $item, array $owner, $topl
-		if (($item['gravity'] != GRAVITY_PARENT) && (Strings::normaliseLink($item["author-link"]) != Strings::normaliseLink($owner["url"]))) {
-			Logger::info("OStatus entry is from author ".$owner["url"]." - not from ".$item["author-link"].". Quitting.");
+		if (($item['gravity'] != GRAVITY_PARENT) && (Strings::normaliseLink($item['author-link']) != Strings::normaliseLink($owner['url']))) {
+			Logger::info("OStatus entry is from author " . $owner['url'] . " - not from " . $item['author-link'] . ". Quitting.");
```
```suggestion
			Logger::info('OStatus entry is from author ' . $owner['url'] . ' - not from ' . $item['author-link'] . '. Quitting.');
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -1680,30 +1734,29 @@ private static function followEntry(DOMDocument $doc, array $item, array $owner,
-		if (($item['gravity'] != GRAVITY_PARENT) && (Strings::normaliseLink($item["author-link"]) != Strings::normaliseLink($owner["url"]))) {
-			Logger::info("OStatus entry is from author ".$owner["url"]." - not from ".$item["author-link"].". Quitting.");
+		if (($item['gravity'] != GRAVITY_PARENT) && (Strings::normaliseLink($item['author-link']) != Strings::normaliseLink($owner['url']))) {
+			Logger::info("OStatus entry is from author " . $owner['url'] . " - not from " . $item['author-link'] . ". Quitting.");
```
```suggestion
			Logger::info('OStatus entry is from author ' . $owner['url'] . ' - not from ' . $item['author-link'] . '. Quitting.');
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -1680,30 +1734,29 @@ private static function followEntry(DOMDocument $doc, array $item, array $owner,
 				$title = BBCode::convertForUriId($item['uri-id'], $item['title'], BBCode::OSTATUS);
 			} else {
-				$title = sprintf("New note by %s", $owner["nick"]);
+				$title = sprintf("New note by %s", $owner['nick']);
```
```suggestion
				$title = sprintf('New note by %s', $owner['nick']);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -1680,30 +1734,29 @@ private static function followEntry(DOMDocument $doc, array $item, array $owner,
 			}
 		} else {
-			$title = sprintf("New comment by %s", $owner["nick"]);
+			$title = sprintf("New comment by %s", $owner['nick']);
```
```suggestion
			$title = sprintf('New comment by %s', $owner['nick']);
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -1767,38 +1819,40 @@ public static function entryHeader(DOMDocument $doc, array $owner, array $item,
 
 		if (!empty($item['title'])) {
-			$body = "[b]".$item['title']."[/b]\n\n".$body;
+			$body = "[b]" . $item['title'] . "[/b]\n\n" . $body;
```
```suggestion
			$body = '[b]' . $item['title'] . "[/b]\n\n" . $body;
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -1880,72 +1935,70 @@ private static function entryFooter(DOMDocument $doc, $entry, array $item, array
-						"href" => $mention]
-				);
+				XML::addElement($doc, $entry, 'link', '', [
+						'rel' => 'mentioned',
+						'ostatus:object-type' => Activity\ObjectType::GROUP,
+						'href' => $mention,
```
```suggestion
					'rel' => 'mentioned',
					'ostatus:object-type' => Activity\ObjectType::GROUP,
					'href' => $mention,
```

MrPetovan commented on src/Protocol/OStatus.php:
```diff
@@ -1880,72 +1935,70 @@ private static function entryFooter(DOMDocument $doc, $entry, array $item, array
-						"href" => $mention]
-				);
+				XML::addElement($doc, $entry, 'link', '', [
+						'rel' => 'mentioned',
+						'ostatus:object-type' => Activity\ObjectType::PERSON,
+						'href' => $mention,
```
```suggestion
					'rel' => 'mentioned',
					'ostatus:object-type' => Activity\ObjectType::PERSON,
					'href' => $mention,
```

Quix0r committed:
Changes:
- added some documentation
- fixed some documentation
- changed more double-quotes to singl

Quix0r committed:
Fixed:
- TypeError: "Argument 1 passed to Friendica\Util\XML::escape() must be of the
  type string, null given, called in src/Util/XML.php line 171

MrPetovan commented on src/Content/Text/BBCode.php:
```diff
@@ -1247,16 +1247,28 @@ private static function removePictureLinksCallback($match)
 
-	private static function expandLinksCallback($match)
+	/**
+	 * Callback: Expands links from given $match array
```
Same comment as the parameter documentation earlier, this is an exact duplicate of the method signature which doesn't provide any more information.

Quix0r replied:
I normally document more than lesser. :)

MrPetovan commented on src/Protocol/DFRN.php:
```diff
@@ -2209,26 +2207,26 @@ private static function processDeletion(DOMXPath $xpath, DOMNode $deletion, arra
 		$item = Post::selectFirst(['id', 'parent', 'contact-id', 'uri-id', 'deleted', 'gravity'], $condition);
 		if (!DBA::isResult($item)) {
-			Logger::info("Item with uri " . $uri . " for user " . $importer['importer_uid'] . " wasn't found.");
+			Logger::info('Item with URI ' . $uri . ' for user ' . $importer['importer_uid'] . ' was not found.');
```
So, the purpose of normalizing quotes to single quotes is to be able to spot the double-quote strings when they are actually needed like in this case where we want to use a quote in a string. We don't have to avoid using double-quotes at all cost, they just need to be used only when they are needed.
